xarray/core/variable.py

```diff
@@ -637,20 +637,19 @@ def __setitem__(self, key, value):
         """
         dims, index_tuple, new_order = self._broadcast_indexes(key)
 
-        if isinstance(value, Variable):
-            value = value.set_dims(dims).data
-        else:  # first broadcast value
+        if not isinstance(value, Variable):
             value = as_compatible_data(value)
             if value.ndim > len(dims):
                 raise ValueError(
                     'shape mismatch: value array of shape %s could not be'
                     'broadcast to indexing result with %s dimensions'
                     % (value.shape, len(dims)))
-
             if value.ndim == 0:
-                value = Variable((), value).set_dims(dims).data
+                value = Variable((), value)
             else:
-                value = Variable(dims[-value.ndim:], value).set_dims(dims).data
+                value = Variable(dims[-value.ndim:], value)
+        # broadcast to become assignable
+        value = value.set_dims(dims).data
```

Comment:
I decided to revert `nputils.NumpyVindexAdapter` to the master version and to make a broadcasting here.